File: D3/kMC/Ex03/Ex03.c
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>

/* ------- VARIABLE DECLARATION... 
   ------------------------------- */

// ::: PHYSICAL PARAMETER
double kbev = 8.61733e-5;
double kb = 1.38e-23;
double na = 6.02214e23;
double pi = 4.0*atan(1.0);
double u = 1.66e-27;
double ang = 1e-10;

// ::: USER VARIABLE

// ::: STRUCTURE 
struct event_type {

  int nevent, nbond, nchem_react;

  int *ptr_i_state,   // array contains Initial state 
      *ptr_f_state;   // array contains Final state 

  double *ptr_f0,          // array contains prefactor rate 
         *ptr_ebarrier,    // array contains activation energies
         *ptr_de,          // array contains difference energy between two miniums
         **ptr_ebond;      // 2D array contains bond energies
};

struct kmc_type {

  int bavard, conv;          // Verbose & convergence
  int period[3], nsites[3];  // Bound periodic condition & number of site 

  char algorithm[50],        // algorithm name: BKL or Gillepsie
       input_file[50],       // input_file name
       input_event[50],      // event file name
       libname[50],          // shared lib name with the path
       init_mod[50];         // Initialization mode: species, random, none 

  int tot_sites,             // Total number of sites
      max_step,              // max step criterium
      sys_dim,               // system dimension, 1, 2 or 3 D
      freq_write,            // write stat frequency
      nprop,                 // number of properties compute in analyse routine
      node_state,            // number of state of one site
      nspec,                 // number of different sepcies
      npressure,             // number of partial pressure
      step;                  // actual step

  double sum_rate,           // Total rate of system 
         rand_rate,          // random rate 
         time,               // actual time
         rand_time,          // random time
         max_time,           // max time criterium
         temp,               // temperature value (Kelvin)
         kt,                 // Energy of temperature
         per100,             // Percentage of 0 state in initial system
         f0,                 // prefactor rate value
         scale;              // scale value 

  int     *ptr_site, *ptr_nneig, *ptr_nevt, *ptr_neig, *ptr_event_site, *ptr_spec;
  double  *ptr_rate, *ptr_prop, *ptr_event_rate, *ptr_pressure, *ptr_masse;

  struct event_type event;
};


extern int *__hidden_table_MOD_h_i_state, *__hidden_table_MOD_h_f_state;
extern double *__hidden_table_MOD_h_ebarrier, *__hidden_table_MOD_h_de, **__hidden_table_MOD_h_ebond, 
              *__hidden_table_MOD_h_f0;


/* ------- FUNCTION DECLARATION...
   ------------------------------- */

void builder_event_type( struct event_type *, int );


/* ------- FUNCTION DEFINITION... 
   ------------------------------ */

char ** parsing( char *string, int len ) {
  int i = 0;
  char **pch;
  pch = malloc( len*sizeof(char) );
  pch[i] = strtok( string, "\t " );
  while ( pch[i] != NULL ) {
   // printf( "sparse %d %s \n",i, pch[i] );
    i++;
    pch[i] = strtok( NULL, "\t " );
  }
  return pch;
}
// ...........................................................
// ......................................................................

double random_number( double min, double max ) {
   return min + (double)rand() / ( (double)(RAND_MAX + 1.0)*( max - min) );
}
// ......................................................................
// .................................................................................
// ......................................................................

void alloc_chem_event_type( struct event_type *event, int nspec ) {
  int int_size = sizeof( int );
  int double_size = sizeof( double );
  int n = event-> nevent + event-> nchem_react*nspec;

  printf( " Enter in EVENT_type constructor... %d %d\n", event->nchem_react, event->nevent );

  __hidden_table_MOD_h_i_state  = malloc( n*int_size );
  __hidden_table_MOD_h_f_state  = malloc( n*int_size );
  __hidden_table_MOD_h_f0       = malloc( n*double_size );
  __hidden_table_MOD_h_ebarrier = malloc( n*double_size );
  __hidden_table_MOD_h_de       = malloc( n*double_size );

  if ( event->nbond != 0 ) {
     __hidden_table_MOD_h_ebond = malloc( event->nbond*sizeof( double* ) );
     for (int i = 0; i < event->nbond; i++ )
        __hidden_table_MOD_h_ebond[i] = malloc( event->nbond*sizeof( double** ) );
  }

  event->ptr_i_state  = __hidden_table_MOD_h_i_state;
  event->ptr_f_state  = __hidden_table_MOD_h_f_state;
  event->ptr_f0       = __hidden_table_MOD_h_f0;
  event->ptr_ebarrier = __hidden_table_MOD_h_ebarrier;
  event->ptr_de       = __hidden_table_MOD_h_de;
  if ( event->nbond != 0 ) 
     event->ptr_ebond    = __hidden_table_MOD_h_ebond;

  printf( " EVENT_type constructor DONE \n" );

}
// ...........................................................
//.................................................................................

void read_event( struct kmc_type *struc ) {

  char string[100], **word;
  char *b = string;
  int i, id, j, nevent, nbond, ibd, jbd, nspec;
  size_t bufsize = 100;
  size_t nword;

  FILE *fp;

  // ::: READ EVENT_FILE :::

  fp = fopen( struc->input_event, "r" );
  if ( fp == NULL ) perror( " PB open input_event...\n" );

  nword = getline(&b, &bufsize, fp );

  while ( strcmp(&string[0],"#") >= 0 ) {

     // SPARSE the line....
     word = parsing( string, bufsize );


     /* ==== read the event ==== */
     if ( !strcmp(word[0],"Number_of_reaction") ) {
        struc->event.nchem_react = atoi( word[1] );
        alloc_chem_event_type( &struc->event, nspec );
        //break;

        for ( i = 0; i < struc->event.nchem_react; i ++ ) {
            nword = getline(&b, &bufsize, fp );
            word = parsing( string, bufsize );

            id = i*struc->event.nchem_react;
            for ( j = 1; j <= struc->nspec; j++ ) {
                struc->event.ptr_i_state[ id + j ] = atoi( word[ j ] );
                struc->event.ptr_f_state[ id + j ] = atoi( word[ struc->nspec + j ] );
                printf( " %d  i %d f %d ", id + j, struc->event.ptr_i_state[ id + j ], struc->event.ptr_f_state[ id + j ] );
            }
            struc->event.ptr_ebarrier[ i ] = atof( word[ 2*struc->nspec + 1 ] );
         }
     } // -------


     /* ==== READ PARTIAL PRESSURE ==== */

     if ( !strcmp( word[0], "Init_species" ) ) {
        
	nspec = atoi( word[1] );
        if ( nspec != struc-> nspec ) {
            printf( "WARNING: Number of species %d is different to input species %d\n", nspec, struc-> nspec );
            exit(1);
         }
         
         for ( i = 0; i < nspec; i++ ) {
             nword = getline(&b, &bufsize, fp );
             word = parsing( string, bufsize );
             struc->ptr_spec[ i ] = atof( word[0] );
             printf( " species: %d %d\n", i, struc->ptr_spec[i] );
         }

     } // -------


     // Read new line...
     nword = getline(&b, &bufsize, fp );


  } // --- END WHILE ----


  fclose( fp );

  srand( time(NULL) );

  //exit(1);

}
// .................................................................................................

void event_rate_calc( struct kmc_type *obj ) {
  
  int is ; 
  double h ;


  for ( is = 0; is < obj->tot_sites; is++ ) {


  } // -- For "is"

}
// .................................................................................................

void choose_event( struct kmc_type *struc, int *isite, int *ievent ) {

  int i, j0, jn, j, nevt, nber;
  double rdn, tmin = 1000.0, t;

   *isite = 0;
   *ievent = 0;


}
// .................................................................................................

void event_applied( struct kmc_type *obj, int *is, int *jn ) {


}
// .................................................................................................

void analyse( struct kmc_type *obj ) {

  int i;

  if ( obj->nprop != obj->nspec ) {
     printf( " Pb nprop %d != nspec %d \n", obj->nprop, obj->nspec );
     return;
  }

  obj->ptr_prop[ 0 ] = 0.0;
  for ( i = 0; i < obj->nspec; i++ ) { 
      obj->ptr_prop[ i ] = obj->ptr_spec[ i ];
  }

}
// .................................................................................................

File: D3/kMC/Ex03/Ex03.f90

  module variables
    use iso_c_binding
    implicit none
    !
    integer( c_int ), dimension(:), pointer :: site, nneig, neig, nevt, spec, event_site
    !
    real( c_double ), dimension(:), pointer :: rate, prop, event_rate
    !
    ! ::: EVENT
    integer( c_int ), dimension(:), pointer :: init_state, final_state
    real( c_double ), dimension(:), pointer :: ebarrier, de, f0
    real( c_double ), dimension(:,:), pointer :: ebond
    !
  end module variables
    !
    !  USER routine:
    !  - subroutine read_event( obj ) 
    !  - subroutine event_rate_calc( obj )
    !  - subroutine choose_event( obj, is, jn )
    !  - subroutine event_applied( obj, is, jn )
    !  - subroutine analyse( obj )
    !
! .................................................................................................
    !
    subroutine alloc_chem_event_type( this, nspec ) bind( C )
      use iso_c_binding
      use derived_types
      use sub_new_types
      use errors
      use variables
      implicit none
      type( event_type )   :: this
      integer( c_int ), intent ( in ) :: nspec
      integer( c_int ) :: n

      print*, " Enter in EVENT_type constructor..."
      !this% nevent = nevt
      n = this% nevent + this% nchem_react*nspec
      print*, " Allocation size :", n, this% nevent, this% nchem_react, nspec
      !
      allocate( h_i_state(n), h_f_state(n), h_ebarrier(n), h_dE(n))
      if ( .not.allocated(h_i_state) .or.  &
           .not.allocated(h_f_state) .or.  &
           .not.allocated(h_ebarrier) .or. &
           .not.allocated(h_de) )          &
         call error( " EVENT_type => CONSTRUCTOR problem " )
      !
      if ( this% nbond /= 0 ) then
         allocate( h_ebond(this% nbond, this% nbond) )
         if ( .not.allocated(h_ebond) )          &
            call error( " EVENT_type => CONSTRUCTOR bond problem " )
      endif
      !
      !
      this% ptr_i_state      = c_loc( h_i_state(1) )
      this% ptr_f_state      = c_loc( h_f_state(1) )
      this% ptr_ebarrier     = c_loc( h_ebarrier(1) )
      this% ptr_de           = c_loc( h_de(1) )
      if ( this% nbond /= 0 )  &
         this% ptr_ebond     = c_loc( h_ebond(1,1) )
      !
      print*, " EVENT_type constructor DONE "
      !
    end subroutine alloc_chem_event_type
! .................................................................................................
!
    subroutine read_event( struc ) bind( C )
      use iso_c_binding
      use derived_types
      use sub_new_types
      use errors
      use variables
      implicit none

      type( KMC_type ), intent( inout ) :: struc
      character (len=50,kind=c_char)    :: string
      integer( c_int )                  :: u0, i, j, id, ios, nspec
      logical                           :: EOF

      !integer( c_int ), dimension(:), pointer :: spec

      character (len=1,kind=c_char)  :: delims
      CHARACTER (len=100,kind=c_char), dimension (50) :: args
      integer( c_int ) :: nargs
      delims = " "
      !
      !
      !  ::: Initialize event parameter
      !
      struc% event% nevent      = 0
      struc% event% nbond       = 0
      struc% event% nchem_react = 0
      !
      !  ::: Lecture of state and rate of each node
      !
      print*, " INPUT_EVENT: ",trim(struc% input_event)
      open( newunit=u0, file=trim(struc% input_event), iostat=ios )
      if ( ios /= 0 ) call error( "input_event does't open!!" )
      !
      EOF = .false.
      do while ( .not.EOF )
         call read_line( u0, string, EOF )
         call parse( trim(string), delims, args, nargs )
         !write (*,*) "Lu :", nargs, (i,trim(args(i)), i=1,nargs)
         !
         if ( args(1) == "Init_species" ) then
            read( args(2), '(i5)' ) nspec
            if ( nspec /= struc% nspec) call error( " PB Number of Init species..." )
            write (*,*) 'Init_species:',trim(args(1)), struc% nspec
            !
            call link_int1_ptr( struc% ptr_spec, spec,        struc% nspec)
            !
            do i = 1,struc% nspec
               read (u0,*) args(1)
               read( args(1), '(i10)' ) spec( i )
               print*, "spec", i, "value", spec(i)
            enddo
         endif
         !
         if ( args(1) == "Number_of_reaction" ) then
            read( args(2), '(i5)' ) struc% event% nchem_react
            call alloc_chem_event_type( struc% event, struc% nspec )
            !exit
            !
            id = struc% event% nchem_react*struc% nspec
            call link_int1_ptr( struc% event% ptr_i_state,   init_state,  id )
            call link_int1_ptr( struc% event% ptr_f_state,   final_state, id )
            call link_real1_ptr( struc% event% ptr_ebarrier, ebarrier,    id )
            !
            do i = 1,struc% event% nchem_react
               id = ( i - 1 )*struc% nspec
               read(u0,*) ios, ( init_state( id + j ), j=1,struc% nspec ),  &
                               ( final_state( id + j ), j=1,struc% nspec ), ebarrier( i )
               write(*,*) id, ( init_state( id + j ), j=1,struc% nspec ),   &
                              ( final_state( id + j ), j=1,struc% nspec ), ebarrier( i )
            enddo
            !
         endif
         !
      enddo
      !
      !
      close( u0 )
      !stop "read_event"
      !
    end subroutine read_event
! .................................................................................................

    subroutine event_rate_calc( obj ) bind( C )
      use iso_c_binding
      use derived_types
      use sub_new_types
      use errors
      use variables
      implicit none
      type( KMC_type ), intent (inout) :: obj
      !
      integer( c_int ) :: is, ir, id, j, n, evt( obj% event% nchem_react )
      real( c_double ) :: kt, h
      !
      n = obj% nspec*obj% event% nchem_react
      call link_int1_ptr( obj% ptr_site,             site,        obj% tot_sites )
      call link_int1_ptr( obj% ptr_nevt,             nevt,        obj% tot_sites )
      call link_real1_ptr( obj% ptr_rate,            rate,        obj% tot_sites )
      call link_int1_ptr( obj% ptr_spec,             spec,        obj% nspec)
      call link_int1_ptr( obj% event% ptr_i_state,   init_state,  n )
      call link_real1_ptr( obj% event% ptr_ebarrier, ebarrier,    n )
      call link_int1_ptr( obj% ptr_event_site,       event_site,  nvois*obj% tot_sites )
      call link_real1_ptr( obj% ptr_event_rate,      event_rate,  nvois*obj% tot_sites )
      !
      !
    end subroutine event_rate_calc
! .................................................................................................

    !real( c_double ) function Chem_combination( ievent ) result ( h )    
 !! 
   ! end function Chem_combination
! .................................................................................................

    subroutine choose_event( obj, is, jn ) bind( C )
      use iso_c_binding
      use derived_types
      use sub_new_types
      use errors
      use variables
      implicit none
      type( KMC_type ) :: obj
      integer( c_int ), intent( out ) :: is, jn
      !

      !
    end subroutine choose_event
! .................................................................................................

    subroutine event_applied( obj, is , jn ) bind( C )
      use iso_c_binding
      use derived_types
      use sub_new_types
      use errors
      use variables
      implicit none
      type( kmc_type ) :: obj
      integer( c_int ), intent( in ) :: is, jn
      integer( c_int ) :: i, id, n
      !
      n = obj% nspec*obj% event% nchem_react
      call link_int1_ptr( obj% ptr_spec,             spec,        obj% nspec )
      call link_int1_ptr( obj% event% ptr_f_state,   final_state, n )
      !
      !
    end subroutine event_applied
! .................................................................................................

    subroutine analyse( obj ) bind( C )
      use iso_c_binding
      use derived_types
      use sub_new_types
      use errors
      use variables
      implicit none
      type( kmc_type ) :: obj
      integer( c_int ) :: i
      !
      call link_int1_ptr( obj% ptr_spec, spec, obj% nspec )
      call link_real1_ptr( obj% ptr_prop, prop, obj% nprop )
      !
      if ( obj% nprop /= obj% nspec ) call error( " ERROR change analyse routine... nprop/=nspec" )
      !
      do i = 1,obj% nprop
         prop( i ) = spec( i ) 
      enddo
      !
    end subroutine
! .................................................................................................

! =================================================================================================










